src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1630,6 +1654,61 @@ private void MakeSeparatorList(ReadOnlySpan<char> separators, ref ValueListBuild
             }
         }
 
+        private void MakeSeparatorListVectorized(ref ValueListBuilder<int> sepListBuilder, char c, char? c2 = null, char? c3 = null)
+        {
+            Vector256<byte> shuffleConstant = Vector256.Create(0x02, 0x06, 0x0A, 0x0E, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+                                                        0xFF, 0xFF, 0xFF, 0xFF, 0x02, 0x06, 0x0A, 0x0E, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
+
+            Vector256<ushort> v1 = Vector256.Create(c);
+            Vector256<ushort>? v2 = c2 is char sep2 ? Vector256.Create(sep2) : (Vector256<ushort>?)null;
+            Vector256<ushort>? v3 = c3 is char sep3 ? Vector256.Create(sep3) : (Vector256<ushort>?)null;
+
+            ref char c0 = ref MemoryMarshal.GetReference(this.AsSpan());
+            int cond = Length - (Length % Vector256<ushort>.Count);
+            int i = 0;
+
+            for (; i < cond; i += Vector256<ushort>.Count)
+            {
+                ref char ci = ref Unsafe.Add(ref c0, i);
+                Vector256<ushort> charVector = Unsafe.As<char, Vector256<ushort>>(ref ci);
+                Vector256<ushort> cmp = Avx2.CompareEqual(charVector, v1);
+
+                if (v2 is Vector256<ushort> vecSep2)
```

Comment:
Thought on eliminating these branches (didn't validate it, don't know if it's worth it).

Instead of checking `v2 is Vector` in each iteration, one could set `v2` to a dummy-value, so that iif `c2` is null the result of `Avx2.Or(Avx2.CompareEqual(charVector, vecSep2), cmp)` won't be changed.

A such dummy-value is `Vector256<byte>.Zero`.
With this value it won't work if the input is all zero though, but otherwise the two branches (in the loop) are eliminated resulting in smaller loop-bodies.

For `v3` analogous.

Reply:
Hmm, good idea! Instead of setting it to `Vector256<byte>.Zero` I could try with the inverse of the other separator char, this way it would even work when `c == '\0'`. I'll give it a benchmark and see if it makes a difference!

Reply:
Okay, upon benchmarking this seems to be significantly slower (-20%) in cases, where there is only a single separator char (which I feel like is probably the most common scenario). I'm leaving this open for now if someone has some other ideas about eliminating the branches.

Reply:
> significantly slower (-20%) in cases, where there is only a single separator char

TBH I can't believe this. Can you show the code?
The branch-predictor does a good job, but such an decrease with branchless operations that can be executed in parallel (instruction level parallelism) is not what I expected.

> the most common scenario

I believe so too. Maybe it is worth it to special case this (when the branchless variant proves to be definitely slower 😉).

Reply:
Unless I am misunderstanding something, were you suggesting the following:

```csharp
if (v2 is Vector256<ushort> vecSep2)
{
    cmp = Avx2.Or(Avx2.CompareEqual(vi, vecSep2), cmp);
}

if (v3 is Vector256<ushort> vecSep3)
{
    cmp = Avx2.Or(Avx2.CompareEqual(vi, vecSep3), cmp);
}
```

to

```csharp
cmp = Avx2.Or(Avx2.CompareEqual(vi, vecSep2), cmp);
cmp = Avx2.Or(Avx2.CompareEqual(vi, vecSep3), cmp);
```
where vecSep2/3 are defaulted to something that wouldn't interfere with the existing cmp value if only one separator was defined?

Reply:
Yep, that's the idea. `vecSep{2,3}` should be hoisted outside the loop, some tweaks (depending on what exactely the JIT produces) and I assume / hope this is faster than the current version with branches.

Reply:
I've ran the numbers again with a proper benchmark and you seem to be correct, it seems to be quite a bit faster, especially in scenarios where TESTZ returns true. Not sure how I managed to get the previous numbers. This is completely dark magic to me though as to why this is faster o.O

For reference:
Numbers: https://gist.github.com/bbartels/9e42086a29afd2a76875e6b4c9e6c8bd
Benchmark: https://gist.github.com/bbartels/5f59bd554e1fe72b9930f7496f97d93a

I will try to adjust the initialization of the other separator vectors, so that they won't interfere with the correctness of the method. Thank you for you input :)

Reply:
👍 I can have a look at the generated code and see if there are some more tweaks. Will do tomorrow...

Reply:
Had a quick look at the generated assembly and see some stack-spills in the loop.
(Maybe that's the reason why you saw a perf-decrease on your first benchmarks).
Will try to eliminate / reduce this spills, and give an update.

Reply:
I just wrote an entire paragraph about how I couldn't figure out what to set the default `vec2/3` registers to, while not manipulating the result. Only to realize seconds later that I can just set it to `v1` in the default case, as it will just `Or` with itself if there are no further separators.
Added the branchless version in https://github.com/dotnet/runtime/pull/38001/commits/3d2a6457df402448df6e5ae570d689fafb31dd26, have a look if this seems good to you!  

Thank you for you work on this, I realize now that there is an even steeper learning curve for these low-level things than I initially imagined 😅

Reply:
Looks quite good (expecte the stack-spills I'm seeing / investigating).

If `\0` isn't a concern, so the default values of the args `c{2,3}` could be set to `\0` to avoid the branches at creating the vectors. This could a have a positive effect on smaller inputs.

With your idea to setting it to `v1` you're on the safe side though.

Reply:
Instead of having nullable arguments, just set the dummy-values in 
https://github.com/dotnet/runtime/blob/140c2ce8182974e674f14d07a73b06c8b67d10b5/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L1577-L1578
as there it's known. You know what I mean?

Reply:
As in calling it like `MakeSeparatorListVectorized(sep0, sep0, sep0)` ?

Reply:
Yep.

```c#
// ...
MakeSeparatorListVectorized(ref sepListBuilder, sep0, sep0, sep0);
// ...
MakeSeparatorListVectorized(ref sepListBuilder, sep0, sep1, sep0);
// ...
MakeSeparatorListVectorized(ref sepListBuilder, sep0, sep1, sep2);
```
Then the checks in `MakeSeparatorListVectorized` for `null` (default arg) can be eliminated.

Reply:
Done https://github.com/dotnet/runtime/pull/38001/commits/2b0b8f8d63b8a3e5af5d2d3e2160c17efc245561